aws_signing_helper/pkcs11_signer.go

```diff
@@ -394,27 +828,78 @@ func GetPKCS11Signer(certIdentifier CertIdentifier, libPkcs11 string, pinPkcs11
 		goto fail
 	}
 
+	// If we found multiple keys, try them until we find the one
+	// that actually matches the cert. More realistically, there
+	// will be only one. Sanity check that it matches the cert.
 	for _, curPrivateKeyHandle := range sessionPrivateKeyObjects {
-		if checkPrivateKeyMatchesCert(module, session, pinPkcs11, curPrivateKeyHandle, certificate, manufacturerId) {
+		// Find the signing algorithm
+		keyAttributes = []*pkcs11.Attribute{
+			pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, 0),
+		}
+		if keyAttributes, err = module.GetAttributeValue(session, curPrivateKeyHandle, keyAttributes); err != nil {
+			continue
+		}
+		keyType, err = bytesToUint(keyAttributes[0].Value)
+		if err != nil {
+			goto fail
+		}
+
+		keyAttributes[0] = pkcs11.NewAttribute(pkcs11.CKA_ALWAYS_AUTHENTICATE, 0)
+		keyAttributes, err = module.GetAttributeValue(session, curPrivateKeyHandle, keyAttributes)
+		if err == nil {
+			alwaysAuth, err = bytesToUint(keyAttributes[0].Value)
+			if err != nil {
+				goto fail
+			}
+		} else {
+			alwaysAuth = 0
+		}
+
+
+		if certificate == nil ||
+			checkPrivateKeyMatchesCert(module, session, keyType, alwaysAuth, pinPkcs11, curPrivateKeyHandle, certificate, manufacturerId) {
 			privateKeyHandle = curPrivateKeyHandle
+			break
 		}
 	}
+
 	if privateKeyHandle == 0 {
+		/*
+		 * "If the key is not found and the original search was by
+		 * CKA_LABEL of the certificate, then repeat the search using
+		 * the CKA_ID of the certificate that was actually found, but
+		 * not requiring a CKA_LABEL match."
+		 *
+		 * http://david.woodhou.se/draft-woodhouse-cert-best-practice.html#rfc.section.8.2
+		 */
+		if (privateKeyId == "" || privateKeyId == certificateId) &&
+			certificate != nil && cert_obj[0].id != nil {
+			_, key_had_label := key_uri.GetPathAttribute("object", false)
+			if key_had_label {
+				key_uri.RemovePathAttribute("object")
+				key_uri.SetPathAttribute("id", escapeAll(cert_obj[0].id))
+				goto retry_search
+			}
+		}
+
 		err = errors.New("unable to find matching private key")
 		goto fail
 	}
 
-	// Find the signing algorithm
-	switch certificate.PublicKey.(type) {
-	case *ecdsa.PublicKey:
+	switch (keyType) {
+	case pkcs11.CKK_EC:
 		signingAlgorithm = aws4_x509_ecdsa_sha256
-	case *rsa.PublicKey:
+	case pkcs11.CKK_RSA:
 		signingAlgorithm = aws4_x509_rsa_sha256
 	default:
 		return nil, "", errors.New("unsupported algorithm")
 	}
 
-	return &PKCS11Signer{certificate, nil, module, session, privateKeyHandle, pinPkcs11}, signingAlgorithm, nil
+	if alwaysAuth == 0 {
```

Comment:
Is the condition flipped here? My understanding of `CKA_ALWAYS_AUTHENTICATE` may be wrong though.